### ISSUE TITLE
Scrolling tab bar

### ISSUE THREAD

**Issue by Akinevz** [herbstluftwm/herbstluftwm#966]
Hiya, could anyone point me towards any resource highlighting how to focus a floating window?

I've recently started using herbstluftwm but I'm currently stuck trying to emulate i3wm's "'[floating]' focus", and "[urgent class='\_\_focused\_\_'] focus". Are there any ways to focus the next (only floating) window using the herbstluftwm client?
![attachment1](https://user-images.githubusercontent.com/1402632/88347522-cda68780-cd42-11ea-9427-3907d094f3fd.png)
![attachment2](https://user-images.githubusercontent.com/1402632/88347625-0d6d6f00-cd43-11ea-9736-e1be36809de8.png)

Apologies if this is not the right place to ask questions, could you point me in the direction of the forum?

**Comment by Akinevz** [comment]
I'm trying to write tetris using herbstluftwm, is that allowed?

**Comment by bklaase** [comment]
maybe something like this:

- make a script
- start by using `cycle_all`
- then check `get_attr clients.focus.floating`
- if it isn't, repeat

you should bookkeep encountered client_id's  to prevent a loop. You could use bash (4+) associative arrays (`declare -A clients`) or something similar.

Too lazy to write it now, but if you have trouble I'll help later.

**Comment by t-wissmann** [comment]
> Hiya, could anyone point me towards any resource highlighting how to focus a floating window?

You can focus floating windows with the directional focus, bound to <kbd>Mod</kbd> + <kbd>h</kbd>/<kbd>j</kbd>/<kbd>k</kbd>/<kbd>l</kbd> per default (for left, down, up, right, the arrow keys are also bound in the default autostart). That means, if you have a floating window focused and want to switch to another floating window that is left of the currently focused window, press <kbd>Mod</kbd>+<kbd>h</kbd>.

> Apologies if this is not the right place to ask questions, could you point me in the direction of the forum?

There is no forum, so asking via github issues is fine for me!

**Comment by bklaase** [comment]
But those will also target tiling windows, which they seem to want to exclude, or not?

**Comment by t-wissmann** [comment]
yes, it will also cover tiling windows if the position of the tiling window is closer.

**Comment by bklaase** [comment]
I think they wanted to filter those out:
> Are there any ways to focus the next (only floating) window using the herbstluftwm client?

So I tested my method by writing a quick script:

```bash
#!/bin/bash

# bookkeeping visited clients
declare -A visitedClients=()
visitedClients["$(herbstclient get_attr clients.focus.winid)"]="true"

# search for floating window
while true ; do
	newWinId="$(herbstclient chain . cycle_all +1 . get_attr clients.focus.winid)"
	[ "${visitedClients[$newWinId]}" = "true" ] && break; # alas, no floating windows found..
	[ "$(herbstclient get_attr clients.$newWinId.floating)" = "true" ] && break;  # found the next floating window!
	
	visitedClients[$newWinId]="true"
done
```

it works, but it is slow. If you have 20 tiling windows in between, you'll see it "searching".

**Comment by Akinevz** [comment]
@bklaase  that is exactly the behaviour I was looking for!

I appreciate the comments. It appears one has to do a linear search manually to locate a floating window programmatically. I was hoping there might be under `herbstclient attr clients.` or `herbstclient attr tags.focused.` but this is an acceptable method, and honestly I'm quite excited by its DIY nature.

From my experience so far, it appears that the floating windows are always together by order. So by starting search from one floating window, one should expect to not get a tiled window using cycle_all, right?

**Comment by bklaase** [comment]
That can, by the nature of cycling, not be true: eventually you would encounter the "last" floating client, and will have to go "through" your tiled clients. In practice the script above would probably be fine for normal use cases, although I cannot anticipate it's effectiveness in HLWM tetris ;) 

I am wondering though, @t-wissmann  , if the behavior of `lock` with regard to cycling can perhaps be altered, to allow for more efficient DIY? Currently cycling within a locked monitor causes `get_attr clients.focus.` to keep pointing to the starting client. If this were not the case, there would not have to be any repaints.

**Comment by Akinevz** [comment]
> That can, by the nature of cycling, not be true: eventually you would encounter the "last" floating client, and will have to go "through" your tiled clients. In practice the script above would probably be fine for normal use cases, although I cannot anticipate its effectiveness in HLWM tetris ;)

Au contraire, that's exactly the behaviour I need! For the actual purpose this is perfect because I only want to cycle floating windows / get to the first floating window. That code can easily be omitted to break on no remaining windows (or focus last tiling window, I'd imagine)

> I cannot anticipate its effectiveness in HLWM tetris ;)

And you'd be surprised what you can do with a few mov-like operations and a dictionary :) 

Also apologies for nonsensical title, it was a late hour when I got this idea. I'll post my code later and mark this as closed for posterity

**Comment by bklaase** [comment]
I misunderstood your question:

> So by starting search from one floating window, one should expect to not get a tiled window using cycle_all, right?

that's right. It will move to the 'next' floating window. If there is none to be found you end up where you started (either a tiled window, or the single floating window)

**Comment by t-wissmann** [comment]
This bash-script sets up a keybinding <kbd>Mod4-d</kbd> that cycles through all floating clients:
```bash
#!/usr/bin/env bash

cycle_only_float=(
  chain
    : lock # avoid window decoration redraws for the next commands
    : and
        # only do something special if this tag is not in
        # 'global' flaoting mode:
        A compare tags.focus.floating = false
        # if we are not on the floating layer
        A compare tags.focus.floating_focused = false
        # then switch to floating layer first
        A attr tags.focus.floating_focused true
    # do the actual cycle on the floating layer
    : cycle_all +1
    : and
        # only do something special if this tag is not in
        # 'global' flaoting mode:
        A compare tags.focus.floating = false
        # check whether we fell out of the floating layer:
        A compare tags.focus.floating_focused = false
        # if we fell out of the floating layer, we switched to tiling
        # and then we are on the first frame.
        A chain
            . cycle_frame -1 # go to the last frame
            . focus_nth 0 # focus the first window in the frame
            . cycle -1 # focus the last window in the frame
            . cycle_all +1 # this jumps back to the floating layer
    : unlock
)

herbstclient keybind Mod4-d "${cycle_only_float[@]}"
```
If the entire tag is in floating mode, then the keybinding acts as an ordinary `cycle_all +1`. This is really hacky and exploits the corner behaviours of a lot of commands, but it works quite robustly :-)

**Comment by Akinevz** [comment]
Hi, I just wanted to say thanks for this code snippet because it's a really
interesting code style for writing config.

Slightly off topic, but how would you recommend a mechanism for checking
whether a keybind exists should work?

**Comment by t-wissmann** [comment]
It seems I've missed the question back then... You can check the existence of a keybinding by running
```
herbstclient list_keybinds
```
and `grep`ing for the keybinding you're interested in.

Note that there are multiple ways to write a keybinding (e.g. `Alt` vs `Mod1` and `-` vs `+`). But `list_keybindings` also prints the keybindings in a normalized way.
